test/e2e/operator/deployment_api.go

```diff
@@ -669,3 +700,25 @@ func validateDeploymentFailure(f *framework.Framework, name string) {
 		return deployment.Status.Phase == api.DeploymentPhaseFailed
 	}, "3m", "5s").Should(BeTrue(), "deployment %q not running", name)
 }
+
+// findOperatorDeployment checks whether there is a PMEM-CSI operator
+// installation in the cluster. An installation is found via its
+// deployment name and namespace.
+func findOperatorDeployment(f *framework.Framework) (*appsv1.Deployment, error) {
+	c, err := deploy.NewCluster(f.ClientSet, f.DynamicClient)
+	if err != nil {
+		return nil, fmt.Errorf("error at create new cluster: %v", err)
+	}
+	d, err := deploy.FindDeployment(c)
+	if err != nil {
+		return nil, fmt.Errorf("find operator deployment: %v", err)
+	}
+
+	framework.Logf("Checking if the operator '%s/%s' running", d.Namespace, d.Name)
+	dep, err := f.ClientSet.AppsV1().Deployments(d.Namespace).Get(context.TODO(), d.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to find operator deployment '%s/%s'", d.Namespace, d.Name)
```

Comment:
This failed in https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-598/lastBuild/testReport/fedora-1_16.operator/API/deployment_driver_image_shall_default_to_operator_image/

It's not immediately obvious to me why it failed, but I haven't looked at the full log of that failure yet.

Reply:
`d.Name` is "operator", which is the high-level name for what should be running in the cluster. It's not the name of the individual objects.

Instead of reimplementing `deploy.WaitForOperator`, let's extend that instead...

I'm working on it.

Reply:
Here's a fix:
```
diff --git a/test/e2e/deploy/deploy.go b/test/e2e/deploy/deploy.go
index f301c252..ab65fbc7 100644
--- a/test/e2e/deploy/deploy.go
+++ b/test/e2e/deploy/deploy.go
@@ -59,15 +59,21 @@ func AddUninstallHook(h UninstallHook) {
 
 // WaitForOperator ensures that the PMEM-CSI operator is ready for use, which is
 // currently defined as the operator pod in Running phase.
-func WaitForOperator(c *Cluster, namespace string) {
+func WaitForOperator(c *Cluster, namespace string) (operator *v1.Pod) {
+       ginkgo.By("waiting for the operator to run")
        // TODO(avalluri): At later point of time we should add readiness support
        // for the operator. Then we can query directoly the operator if its ready.
        // As interm solution we are just checking Pod.Status.
        gomega.Eventually(func() bool {
                pod, err := c.GetAppInstance("pmem-csi-operator", "", namespace)
-               return err == nil && pod.Status.Phase == v1.PodRunning
+               if err == nil && pod.Status.Phase == v1.PodRunning {
+                       operator = pod
+                       return true
+               }
+               return false
        }, "5m", "2s").Should(gomega.BeTrue(), "operator not running in namespace %s", namespace)
-       ginkgo.By("Operator is ready!")
+
+       return
 }
 
 // WaitForPMEMDriver ensures that the PMEM-CSI driver is ready for use, which is
diff --git a/test/e2e/operator/deployment_api.go b/test/e2e/operator/deployment_api.go
index 6c7aa746..feed6b11 100644
--- a/test/e2e/operator/deployment_api.go
+++ b/test/e2e/operator/deployment_api.go
@@ -14,7 +14,6 @@ import (
        pmemtls "github.com/intel/pmem-csi/pkg/pmem-csi-operator/pmem-tls"
        "github.com/intel/pmem-csi/test/e2e/deploy"
 
-       appsv1 "k8s.io/api/apps/v1"
        corev1 "k8s.io/api/core/v1"
        storagev1beta1 "k8s.io/api/storage/v1beta1"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -111,11 +110,11 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
                        deployment, err := toDeployment(dep)
                        Expect(err).ShouldNot(HaveOccurred(), "unstructured to deployment conversion")
 
-                       operatorDep, err := findOperatorDeployment(f)
+                       operatorPod, err := findOperatorDeployment(f)
                        Expect(err).ShouldNot(HaveOccurred(), "find operator deployment")
 
                        // operator image should be the driver image
-                       deployment.Spec.Image = operatorDep.Spec.Template.Spec.Containers[0].Image
+                       deployment.Spec.Image = operatorPod.Spec.Containers[0].Image
                        deploy.CreateDeploymentCR(f, dep)
                        defer deploy.DeleteDeploymentCR(f, deployment.Name)
 
@@ -704,7 +703,7 @@ func validateDeploymentFailure(f *framework.Framework, name string) {
 // findOperatorDeployment checks whether there is a PMEM-CSI operator
 // installation in the cluster. An installation is found via its
 // deployment name and namespace.
-func findOperatorDeployment(f *framework.Framework) (*appsv1.Deployment, error) {
+func findOperatorDeployment(f *framework.Framework) (*corev1.Pod, error) {
        c, err := deploy.NewCluster(f.ClientSet, f.DynamicClient)
        if err != nil {
                return nil, fmt.Errorf("error at create new cluster: %v", err)
@@ -714,11 +713,5 @@ func findOperatorDeployment(f *framework.Framework) (*appsv1.Deployment, error)
                return nil, fmt.Errorf("find operator deployment: %v", err)
        }
 
-       framework.Logf("Checking if the operator '%s/%s' running", d.Namespace, d.Name)
-       dep, err := f.ClientSet.AppsV1().Deployments(d.Namespace).Get(context.TODO(), d.Name, metav1.GetOptions{})
-       if err != nil {
-               return nil, fmt.Errorf("Failed to find operator deployment '%s/%s'", d.Namespace, d.Name)
-       }
-
-       return dep, nil
+       return deploy.WaitForOperator(c, d.Namespace), nil
 }
```

Reply:
> Here's a fix:
> diff --git a/test/e2e/deploy/deploy.go b/test/e2e/deploy/deploy.go
> index f301c252..ab65fbc7 100644
> --- a/test/e2e/deploy/deploy.go
> +++ b/test/e2e/deploy/deploy.go
> @@ -59,15 +59,21 @@ func AddUninstallHook(h UninstallHook) {
>  
>  // WaitForOperator ensures that the PMEM-CSI operator is ready for use, which is
>  // currently defined as the operator pod in Running phase.
> -func WaitForOperator(c *Cluster, namespace string) {
> +func WaitForOperator(c *Cluster, namespace string) (operator *v1.Pod) {
> +       ginkgo.By("waiting for the operator to run")
>         // TODO(avalluri): At later point of time we should add readiness support
>         // for the operator. Then we can query directoly the operator if its ready.
>         // As interm solution we are just checking Pod.Status.
>         gomega.Eventually(func() bool {
>                 pod, err := c.GetAppInstance("pmem-csi-operator", "", namespace)
> -               return err == nil && pod.Status.Phase == v1.PodRunning
> +               if err == nil && pod.Status.Phase == v1.PodRunning {
> +                       operator = pod
> +                       return true
> +               }
> +               return false
>         }, "5m", "2s").Should(gomega.BeTrue(), "operator not running in namespace %s", namespace)
> -       ginkgo.By("Operator is ready!")
> +
> +       return
>  }
>  
>  // WaitForPMEMDriver ensures that the PMEM-CSI driver is ready for use, which is
> diff --git a/test/e2e/operator/deployment_api.go b/test/e2e/operator/deployment_api.go
> index 6c7aa746..feed6b11 100644
> --- a/test/e2e/operator/deployment_api.go
> +++ b/test/e2e/operator/deployment_api.go
> @@ -14,7 +14,6 @@ import (
>         pmemtls "github.com/intel/pmem-csi/pkg/pmem-csi-operator/pmem-tls"
>         "github.com/intel/pmem-csi/test/e2e/deploy"
>  
> -       appsv1 "k8s.io/api/apps/v1"
>         corev1 "k8s.io/api/core/v1"
>         storagev1beta1 "k8s.io/api/storage/v1beta1"
>         metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
> @@ -111,11 +110,11 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
>                         deployment, err := toDeployment(dep)
>                         Expect(err).ShouldNot(HaveOccurred(), "unstructured to deployment conversion")
>  
> -                       operatorDep, err := findOperatorDeployment(f)
> +                       operatorPod, err := findOperatorDeployment(f)
>                         Expect(err).ShouldNot(HaveOccurred(), "find operator deployment")
>  
>                         // operator image should be the driver image
> -                       deployment.Spec.Image = operatorDep.Spec.Template.Spec.Containers[0].Image
> +                       deployment.Spec.Image = operatorPod.Spec.Containers[0].Image
>                         deploy.CreateDeploymentCR(f, dep)
>                         defer deploy.DeleteDeploymentCR(f, deployment.Name)
>  
> @@ -704,7 +703,7 @@ func validateDeploymentFailure(f *framework.Framework, name string) {
>  // findOperatorDeployment checks whether there is a PMEM-CSI operator
>  // installation in the cluster. An installation is found via its
>  // deployment name and namespace.
> -func findOperatorDeployment(f *framework.Framework) (*appsv1.Deployment, error) {
> +func findOperatorDeployment(f *framework.Framework) (*corev1.Pod, error) {
>         c, err := deploy.NewCluster(f.ClientSet, f.DynamicClient)
>         if err != nil {
>                 return nil, fmt.Errorf("error at create new cluster: %v", err)
> @@ -714,11 +713,5 @@ func findOperatorDeployment(f *framework.Framework) (*appsv1.Deployment, error)
>                 return nil, fmt.Errorf("find operator deployment: %v", err)
>         }
>  
> -       framework.Logf("Checking if the operator '%s/%s' running", d.Namespace, d.Name)
> -       dep, err := f.ClientSet.AppsV1().Deployments(d.Namespace).Get(context.TODO(), d.Name, metav1.GetOptions{})
> -       if err != nil {
> -               return nil, fmt.Errorf("Failed to find operator deployment '%s/%s'", d.Namespace, d.Name)
> -       }
> -
> -       return dep, nil
> +       return deploy.WaitForOperator(c, d.Namespace), nil
>  }

I took this patch.